wagtail/whitelist.py

```diff
@@ -99,7 +101,7 @@ class Whitelister:
     def clean(self, html):
         """Clean up an HTML string to contain just the allowed elements /
         attributes"""
-        doc = BeautifulSoup(html, "html5lib")
+        doc = BeautifulSoup(html, BEAUTIFUL_SOUP_FEATURES)
```

Comment:
Flagging up a possible footgun here - we need to be sure that changing the parser doesn't have any side-effects on the HTML output returned from BeautifulSoup.

This whitelister module is at the heart of a big pile of mostly-vestigial code that - I believe - only comes into play when an alternative rich text editor such as [wagtail-hallo](https://github.com/wagtail-nest/wagtail-hallo) is in use. Unlike Draftail, which maintains its own JSON-ish data structure (aka "contentstate") which is posted as part of the form submission and transformed to [Wagtail's pseudo-HTML rich text database format](https://docs.wagtail.org/en/stable/extending/rich_text_internals.html#data-format), this family of rich text editors work directly at the HTML level, which means the equivalent transformation is from untrusted user-supplied HTML to Wagtail's trusted pseudo-HTML subset - i.e. whitelisting.

This pseudo-HTML format has quite finnicky requirements around things like self-closing `/>` tags and `&foo;` entities, because our front-end template tags transform it to real HTML using regexps (yes, yes, I know, but having a tightly controlled source format is exactly what makes that feasible).

Back when this code was in regular use, we had to keep close tabs on BeautifulSoup's output to make sure that it conformed to those requirements - see #3879 for example. Consequently, I'm a little paranoid about innocent-seeming changes to the BeautifulSoup config. It may be that the HTML-generation side of bs4 is entirely decoupled from the parser in use - in which case my worries are unfounded - but it's definitely something to check.

Reply:
From looking into the CI failures, your fears are very much well-founded - swapping out the parser can lead to differences in the rendered output. For example, the current issues are cased by the fact `html5lib` considered the content of `script` tags as part of the plain text of HTML. `html.parser` doesn't. 

Besides this issue, the rest of the test suite passes identically with either implementation.

However, given we've already proved it's a little fiddly, I'm now inclined to make it configurable instead (moving the default to `html.parser` to avoid the dependency). This way niche issues can be avoided by switching back to `html5lib`, and people with specific performance needs can use `lxml` if they wish.

Reply:
I ended up going a different direction and always using `html.parser`. A consistent backend is probably more important than the performance of `lxml` - especially as `html.parser` is already faster than `html5lib`.